src/utils/allocator/page_slab_memory_resource.hpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <cstddef>
+#include <cstdlib>
+#include <memory_resource>
+
+namespace memgraph::utils {
+
+struct PageSlabMemoryResource : std::pmr::memory_resource {
+  static constexpr std::size_t PAGE_SIZE = 4096;
+  PageSlabMemoryResource() = default;
+
+  PageSlabMemoryResource(PageSlabMemoryResource const &) = delete;
+  PageSlabMemoryResource &operator=(PageSlabMemoryResource const &) = delete;
+
+  PageSlabMemoryResource(PageSlabMemoryResource &&other) noexcept {
+    std::swap(pages, other.pages);
+    std::swap(ptr, other.ptr);
+    std::swap(space, other.space);
+  }
+  PageSlabMemoryResource &operator=(PageSlabMemoryResource &&other) noexcept {
+    if (this == &other) return *this;
+    std::swap(*this, other);
+    return *this;
+  }
+
+  ~PageSlabMemoryResource() override {
+    auto current = pages;
+    while (current) {
+      auto next = current->next;
+      std::free(current);
+      current = next;
+    }
+  }
+
+ private:
+  struct header {
+    explicit header(header *next) : next(next) {}
+    header *next = nullptr;
+  };
+
+  constexpr static size_t alignSize(size_t size, size_t alignment) { return (size + alignment - 1) & ~(alignment - 1); }
+
+  void *do_allocate(size_t bytes, size_t alignment) final {
+    // 1. could this fit inside a page slab?
+    auto earlest_slab_position = alignSize(sizeof(header), alignment);
+    auto max_slab_capacity = PAGE_SIZE - earlest_slab_position;
+    if (max_slab_capacity < bytes) [[unlikely]] {
+      auto required_bytes = bytes + earlest_slab_position;
+      auto *newmem = reinterpret_cast<header *>(std::aligned_alloc(alignment, required_bytes));
+      if (!newmem) throw std::bad_alloc{};
+      // add to the allocation list
+      pages = std::construct_at<header>(newmem, pages);
+      return pages + earlest_slab_position;
+    }
+
+    // 2. can it fit in existing slab?
+    if (!std::align(alignment, bytes, ptr, space)) {
+      auto *newmem = reinterpret_cast<header *>(std::aligned_alloc(PAGE_SIZE, PAGE_SIZE));
+      if (!newmem) throw std::bad_alloc{};
+      pages = std::construct_at<header>(newmem, pages);
+      ptr = reinterpret_cast<std::byte *>(pages) + sizeof(header);
+      space = PAGE_SIZE - sizeof(header);
+      std::align(alignment, bytes, ptr, space);
+    }
+
+    // 3. use current slab
+    void *res = ptr;
+    ptr = reinterpret_cast<std::byte *>(ptr) + bytes;
+    space -= bytes;
+    return res;
```

Comment:
Code is not supper clear. I understand we already aligned this, but it's not super clear on first look.